src/v/raft/group_configuration.cc

```diff
@@ -327,194 +679,172 @@ void group_configuration::replace(
 }
 
-bool group_configuration::maybe_demote_removed_voters() {
-    vassert(
-      _old,
-      "can not demote removed voters as configuration is of simple type - {}",
-      *this);
+void configuration_change_strategy_v4::abort_configuration_change(
+  model::revision_id rev) {
+    // collect all node ids
+    absl::flat_hash_set<vnode> all_node_ids;
 
-    // no voters are present, do nothing
-    if (_old->voters.empty()) {
-        return false;
-    }
-    // if voter was removed, make it a learner
-    auto it = std::stable_partition(
-      _old->voters.begin(), _old->voters.end(), [this](const vnode& v) {
-          return _current.contains(v);
-      });
+    _cfg.for_each_learner(
+      [&all_node_ids](const vnode& learner) { all_node_ids.insert(learner); });
 
-    // nothing to remove
-    if (std::distance(it, _old->voters.end()) == 0) {
-        return false;
-    }
+    _cfg.for_each_voter(
+      [&all_node_ids](const vnode& voter) { all_node_ids.insert(voter); });
 
-    std::move(it, _old->voters.end(), std::back_inserter(_old->learners));
-    _old->voters.erase(it, _old->voters.end());
+    // clear the configuration
+    _cfg._current.voters.clear();
+    _cfg._current.learners.clear();
+    _cfg._old.reset();
 
-    return true;
+    // rebuild configuration
+
+    for (auto& vn : all_node_ids) {
+        if (!_cfg._configuration_update->is_to_add(vn)) {
+            _cfg._current.voters.push_back(vn);
+        }
+    }
+    std::erase_if(_cfg._brokers, [this](const model::broker& br) {
+        return !std::any_of(
+          _cfg._current.voters.begin(),
+          _cfg._current.voters.end(),
+          [id = br.id()](vnode vn) { return vn.id() == id; });
+    });
+    // finally reset configuration update and set the revision
+    _cfg._configuration_update.reset();
+    _cfg._revision = rev;
 }
 
-void group_configuration::abort_configuration_change(model::revision_id rev) {
-    vassert(
-      _old,
-      "can not abort configuration change if it is of simple type - "
-      "{}",
-      *this);
+void configuration_change_strategy_v4::cancel_configuration_change(
+  model::revision_id rev) {
+    switch (_cfg.get_state()) {
+    case configuration_state::simple:
+        vassert(
+          false,
+          "can not cancel, configuration change is not in progress - {}",
+          _cfg);
+    case configuration_state::transitional:
+        cancel_update_in_transitional_state();
+        break;
+    case configuration_state::joint:
+        cancel_update_in_joint_state();
+        break;
+    }
+    _cfg._revision = rev;
+}
 
-    absl::flat_hash_set<model::node_id> physical_node_ids;
+void configuration_change_strategy_v4::cancel_update_in_transitional_state() {
+    /**
+     * in transitional state the only change applied to configuration are nodes
+     * which are added to the replica set. We may simply remove learners but for
+     * the learners that were already promoted to voter we must use joint
```

Comment:
looks like we do a transitional -> transitional step here (and removing learners). So when will we move to the joint configuration? On the next iteration? Can we simply move directly to the joint config (or simple config if there were no voter promotions)?

Reply:
Answer is the same for both of the questions. It would be possible to optimize the transitions, i wanted to make it consistent and simple to reason about. This is why i decided to perform the state transition after the configuration is committed. This way we do not need an additional piece of logic checking special cases here. Do you think it would be valuable to add those here ?

Reply:
Optimizations look easy code-wise, so probably a good idea :) And makes it closer to the diagram so possible transitions are easier to understand.